sql/sqlite/driver.go

```diff
@@ -181,19 +186,90 @@ func openTx(ctx context.Context, db *sql.DB, opts *sql.TxOptions) (*sqlclient.Tx
 	if err != nil {
 		return nil, err
 	}
+	before, err := violations(ctx, tx)
+	if err != nil {
+		return nil, err
+	}
 	return &sqlclient.Tx{
 		Tx: tx,
-		Close: func() error {
-			if on.Bool {
-				if _, err := db.ExecContext(ctx, "PRAGMA foreign_keys = on"); err != nil {
-					return fmt.Errorf("sql/sqlite: set 'foreign_keys = on': %w", err)
+		Commit: func() error {
+			after, err := violations(ctx, tx)
+			if err != nil {
+				if err2 := tx.Rollback(); err2 != nil {
+					err = fmt.Errorf("%v: %w", err2, err)
+				}
+				return enableFK(ctx, db, on.Bool, err)
+			}
+			if vs := newViolations(before, after); len(vs) > 0 {
+				err := fmt.Errorf("sql/sqlite: foreign key mismatch: %+v", vs)
+				if err2 := tx.Rollback(); err2 != nil {
+					err = fmt.Errorf("%v: %w", err2, err)
 				}
+				return enableFK(ctx, db, on.Bool, err)
 			}
-			return nil
+			return enableFK(ctx, db, on.Bool, tx.Commit())
+		},
+		Rollback: func() error {
+			return enableFK(ctx, db, on.Bool, tx.Rollback())
 		},
 	}, nil
 }
 
+func enableFK(ctx context.Context, db *sql.DB, do bool, err error) error {
+	if !do {
+		return err
+	}
+	// Re-enable foreign key checks if they were enabled before.
+	if _, err2 := db.ExecContext(ctx, "PRAGMA foreign_keys = on"); err2 != nil {
+		err2 = fmt.Errorf("sql/sqlite: set 'foreign_keys = on': %w", err)
+		if err != nil {
+			return fmt.Errorf("%v: %w", err2, err)
+		}
+		return err2
+	}
+	return nil
```

Comment:
You ignore the case where the above command succeeded, but you still have an error to return.